### PR TITLE
fix: support resource logs and exec #9783

### LINKED ISSUE #9783: Argo CD 2.4.0: RBAC resources exec and logs not known to the projects REST API
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

Hey there! 
We started upgrading our Argo CD instances to 2.4.0. To give permissions to our developers, we create a role on each project in ArgoCD allowing the assigned groups to only do certain actions with their apps. As 2.4.0 introduces two new RBAC resources `logs` and `exec` we started adding them to our roles too. However we encounter issues when creating new Argo CD projects using the REST API. It complains that the RBAC resources `exec` and `logs` are not found.

The exact error message:

```json
{"error":"invalid policy rule 'p, proj:guestbook:guestbook, exec, create, guestbook/guestbook, allow': project resource must be: 'applications', 'repositories' or 'clusters', not 'exec'","code":3,"message":"invalid policy rule 'p, proj:guestbook:guestbook, exec, create, guestbook/guestbook, allow': project resource must be: 'applications', 'repositories' or 'clusters', not 'exec'"}
```

This must either be a bug, or a misunderstanding on our side as the resources `exec` and `logs` are not meant to be set at project-level.

**To Reproduce**

Do an API Call to Argo CD REST API using the following command:

```bash
curl -X POST -d @guestbook.json $ARGOCD_SERVER/api/v1/projects -H "Authorization: Bearer $ARGOCD_TOKEN"  -H 'Content-Type: application/json' -v 
```

where `guestbook.json` contains the following:
```json
{
  "project": {
    "metadata": {
      "name": "guestbook",
      "namespace": "argocd"
    },
    "spec": {
      "clusterResourceBlacklist": [
        {
          "group": "none",
          "kind": "none"
        }
      ],
      "description": "ArgoCD Project for guestbook",
      "destinations": [
        {
          "namespace": "guestbook",
          "server": "https://kubernetes.default.svc"
        }
      ],
      "roles": [
        {
          "groups": [
            "mygroup",
            "anothergroup"
          ],
          "name": "guestbook",
          "policies": [
            "p, proj:guestbook:guestbook, applications, action/*, guestbook/guestbook, allow",
            "p, proj:guestbook:guestbook, applications, delete, guestbook/guestbook, allow",
            "p, proj:guestbook:guestbook, applications, get, guestbook/guestbook, allow",
            "p, proj:guestbook:guestbook, applications, sync, guestbook/guestbook, allow",
            "p, proj:guestbook:guestbook, exec, create, guestbook/guestbook, allow",
            "p, proj:guestbook:guestbook, logs, get, guestbook/guestbook, allow"
          ]
        }
      ],
      "sourceRepos": [
        "https://github.com/liechtinat/guestbook.git"
      ]
    }
  },
  "upsert": true
}
```

This results in the following error message:

```json
{"error":"invalid policy rule 'p, proj:guestbook:guestbook, exec, create, guestbook/guestbook, allow': project resource must be: 'applications', 'repositories' or 'clusters', not 'exec'","code":3,"message":"invalid policy rule 'p, proj:guestbook:guestbook, exec, create, guestbook/guestbook, allow': project resource must be: 'applications', 'repositories' or 'clusters', not 'exec'"}
```

**Expected behavior**

The Project should be created without any error as `exec` and `logs` are valid RBAC resources since v2.4.0.

**Version**

```shell
v2.4.0+91aefab
```

**Logs**

```
time="2022-06-24T13:07:17Z" level=info msg="received unary call /project.ProjectService/Create" grpc.method=Create grpc.request.claims="{\"iat\":1651649987,\"iss\":\"argocd\",\"jti\":\"dce64e3a-1cd7-46a4-b12b-8a00d2fdcd49\",\"nbf\":1651649987,\"sub\":\"username\"}" grpc.request.content="%!v(PANIC=String method: reflect.Value.Bytes of non-byte slice)" grpc.service=project.ProjectService grpc.start_time="2022-06-24T13:07:17Z" span.kind=server system=grpc
time="2022-06-24T13:07:17Z" level=info msg="finished unary call with code InvalidArgument" error="rpc error: code = InvalidArgument desc = invalid policy rule 'p, proj:guestbook1:guestbook1, exec, create, guestbook1/guestbook1, allow': project resource must be: 'applications', 'repositories' or 'clusters', not 'exec'" grpc.code=InvalidArgument grpc.method=Create grpc.service=project.ProjectService grpc.start_time="2022-06-24T13:07:17Z" grpc.time_ms=12.376 span.kind=server system=grpc
```

### PR DESCRIPTION
**Pull request by muma378**
Fixes #9783

Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: pkg/apis/application/v1alpha1/types.go [hunk 1]
```diff
@@ -1555,7 +1555,7 @@ func validatePolicy(proj string, role string, policy string) error {
 	}
 	// resource
 	resource := strings.Trim(policyComponents[2], " ")
-	if resource != "applications" && resource != "repositories" && resource != "clusters" {
+	if resource != "applications" && resource != "repositories" && resource != "clusters" && resource != "exec" && resource != "logs" {
```
Comment (jessesuen) [inline pkg/apis/application/v1alpha1/types.go:1558]: While you're here, do you mind cleaning up and copying what we did for validActions:

```golang
var validActions = map[string]bool{
	"get":      true,
	"create":   true,
	"update":   true,
	"delete":   true,
	"sync":     true,
	"override": true,
	"*":        true,
}
```

Comment (muma378) [reply pkg/apis/application/v1alpha1/types.go:1558]: for sure, should I open a new PR or just commit to this one?

**Review by crenshaw-dev — APPROVED** [review]
lgtm, thanks!

**Comment by crenshaw-dev** [comment]
Cherry-picked onto release-2.4 for 2.4.4.

**Pull request merged**